src/expressions/functions/builtInFunctions.sequences.js

```diff
@@ -469,6 +474,133 @@ function fnFilter (dynamicContext, executionParameters, staticContext, sequence,
 	});
 }
 
+function fnForEach (dynamicContext, executionParameters, staticContext, sequence, callbackSequence) {
+	if (sequence.isEmpty()) {
+		return sequence;
+	}
+
+	/**
+	 * @type {FunctionValue}
+	 */
+	const callbackFn = callbackSequence.first();
+	const callbackArgumentTypes = callbackFn.getArgumentTypes();
+	if (callbackArgumentTypes.length !== 1) {
+		throw new Error(`XPTY0004: signature of function passed to fn:for-each is incompatible.`);
+	}
+
+	const sequences = sequence.map(item => {
+		// Transform argument
+		const transformedArgument = transformArgument(
+			callbackArgumentTypes[0],
+			Sequence.singleton(item),
+			executionParameters,
+			'fn:for-each');
+		return callbackFn.value.call(
+			undefined,
+			dynamicContext,
+			executionParameters,
+			staticContext,
+			transformedArgument);
+	});
+
+	const outerIterator = sequences.value();
+	let innerIterator;
+	return new Sequence({
+		next: () => {
+			while (true) {
+				if (!innerIterator) {
+					const nextSequence = outerIterator.next();
+					// TODO: Consider handling the ready state
```

Comment:
Yup, please handle it by returning the nextSequence variable if the ready property is set to false.